### PR TITLE
[TE] yyuen/entity mapping custom input

### PR DESCRIPTION
**Pull request by justYves**
### What's new: 
- added the following entity mapping: from {metric} to {lixtag|service|dimension|dataset|}
- added validation for simple mapping (urns)
- added type ahead for datasets and services
- added test for both simple and advanced mapping
- abstracted entity mapping apis into its own file

### Screenshot: 
screen shot 2018-03-30 at 9 46 20 pm

### Tests: 
All test passing (287/287) + visual check

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js [hunk 1]
```diff
@@ -431,31 +507,58 @@ export default Component.extend({
...
-      const taskInstance = task.perform(metrics);
-      this.set('mostRecentSearch', taskInstance);
+    onSearch(searchString) {
+      const {
+        lastSearch,
+        searchEntitiesTask: task
+      } = getProperties(this, 'lastSearch', 'searchEntitiesTask');
+
+      searchString = searchString.length ? searchString : lastSearch;
+      const taskInstance = task.perform(searchString);
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: Unfortunately ember concurrency's task are currently not compatible with await/async. I believe @newsummit found that out a few weeks ago

File: thirdeye/thirdeye-frontend/app/helpers/extract-dropdown-value.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { helper } from '@ember/component/helper';
+
+/**
+ * Template helper intended to be used with the powerselec component
+ * this extracts the item's value dynamically
+ * @param {Object}        element     - object found by power select
+ * @param {String}        propertyKey - key property to extract
+ * @return {String}         - value to be extracted
+ */
+export function extractDropDownValue([element, propertyKey]) {
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/helpers/extract-dropdown-value.js]: Yes, it's a template helper, so it's dasherized in the template https://github.com/linkedin/pinot/pull/2689/files#diff-fa2088e19d4061108d3a57d8e9aeefbaR49

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/template.hbs [hunk 3]
```diff
@@ -19,54 +19,58 @@
...
+          type="text"
+          id="simple-entity-selection"
+          class="ember-power-select-trigger te-modal__filter-select te-modal__filter-select--flushed te-modal__input"
+          value=urn
+          key-press="onKeyPress"
+        }}
       {{/if}}
-      <a class="thirdeye-link thirdeye-link--secondary {{if mappingExists 'thirdeye-link--disabled'}}" {{action "onAddFilter"}}>
+
+      <a class="thirdeye-link thirdeye-link--secondary {{unless canAddMapping 'thirdeye-link--disabled'}}" {{action "onAddFilter"}}>
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/template.hbs]: Good to know. Thanks!! will fix

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: Hmm. That's odd. Cause i'm using in my codes just fine. 
Ok, you can leave it for now. Thanks!

```const anomalyMapping = await this.get('_getAnomalyMapping').perform(model);```

File: thirdeye/thirdeye-frontend/app/helpers/extract-dropdown-value.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/helpers/extract-dropdown-value.js]: Ah ok. I wasn't able to find it earlier. Thanks!

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js [hunk 4]
```diff
@@ -343,74 +333,161 @@ export default Component.extend({
...
+    const url = this.makeUrlString(entities);
+    this.builUrnToId(entities);
 
+    if (!url) {
+      return;
+    }
+    const relatedEntities = await fetch(url).then(checkStatus);
+
+    // merges createBy Props
+    relatedEntities.map((item) => {
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: This is actually constructing a new array and setting it here https://github.com/linkedin/pinot/pull/2689/files/c21017e56023f0b5e60ffb788c9ef6f532860ee9#diff-8ed4175cf53b2ef4678cc9b4ee73238fR436

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js [hunk 5]
```diff
@@ -431,31 +507,58 @@ export default Component.extend({
...
-      const task = get(this, 'searchEntities');
-      const taskInstance = task.perform(metrics);
-      this.set('mostRecentSearch', taskInstance);
+    onSearch(searchString) {
+      const {
+        lastSearch,
+        searchEntitiesTask: task
+      } = getProperties(this, 'lastSearch', 'searchEntitiesTask');
+
+      searchString = searchString.length ? searchString : lastSearch;
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: oh Good catch. this should be `lastSearchTerm` consistently. I used to cache the task for a potential `.cancel()`, but this is handled more elegantly with a `.restartable()`

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js [hunk 6]
```diff
@@ -343,74 +333,161 @@ export default Component.extend({
...
     return fetch(url, headers)
       .then(checkStatus);
-  }),
+  }).restartable(),
 
-  actions: {
+  async init() {
+    this._super(...arguments);
+
+    const datasets = await fetch(entityMappingApi.getDatasetsUrl).then(checkStatus);
```
[the remark this answers is not in the document]
Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: I don't think I'll have bandwidth to update this specific piece. @aaronucsd Would mind fixing in a follow up pr?

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: wouldn't you need to set the `relatedEntitiesUpdated = relatedEntities.map((item) => {` back? As it's not self referencing with the mutated new array.

set(this, '_relatedEntities', relatedEntitiesUpdated);



```
var array1 = [1, 4, 9, 16];

// pass a function to map
const map1 = array1.map(x => x * 2);

console.log(map1, array1);
// expected output: Array [2, 8, 18, 32], Array [1, 4, 9, 16]
```

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: Yes. Please leave it! Thanks.

File: thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaronucsd) [reply thirdeye/thirdeye-frontend/app/pods/components/modals/entity-mapping-modal/component.js]: Ok, just making sure we're not introducing a bug. Thanks again Yves. All other minors please leave it and i can followup on another PR.

**Pull request merged**